README
README file for GraphViz2::Marpa.

See also: CHANGES and Changelog.ini.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip GraphViz2-Marpa-1.02.tgz
shell>tar mxvf GraphViz2-Marpa-1.02.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Marpa.pm through you favourite pod2html translator.

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip GraphViz2-Marpa-1.02.zip
shell>ppm install --location=. GraphViz2-Marpa
shell>del GraphViz2-Marpa-1.02.ppd
shell>del PPM-GraphViz2-Marpa-1.02.tar.gz


CHANGES
Revision history for Perl extension GraphViz2::Marpa.

1.09  Fri May 31 09:57:00 2013
	- Re-write the code in Build.PL and Makefile.PL which checks for Graphviz (dot) being installed,
		since the previous code, using a pipe, was failing on some versions of Windows.
	- Assume Config.pm is installed, and hence remove it from the pre-reqs.
		This also stops a warning message generated because Config's version # is undef.
	- Likewise assume File::Spec and File::Temp are installed, and 'recent enough'.
		This is because the new code uses these 3 modules before specifying the pre-reqs.
	- Bump the pre-req for Perl from 5.10.0 to 5.14, since we 'use feature qw/unicode_strings/.
	- Re-write the code in Build.PL and Makefile.PL which checks for Perl being 'recent enough',
		since the previous code, using a pre-req of "perl => '5.10.0'" generates a warning message
		when using Makefile.PL (for my current Perl V 5.14.2). Now we analyze $Config{version}.
	- One CPAN Tester found tests failing even though Graphviz was installed.
		The problem was that the Times font was missing. The new code should fail during 'perl Build.PL',
		or 'perl Makefile.PL', rather than during testing, which is good.
	- Rewrite the use of given/when in GraphViz2::Marpa::Renderer::GraphViz2 as a chain of 'if'
		statements, in preparation for the removal of 'use feature qw/switch/' from Perl.

1.08  Mon May 27 09:45:00 2013
	- Add data/56.gv, which is graphs/directed/biological.gv from Graphviz V 2.29.
	- Add data/57.gv, which is graphs/directed/arrows.gv from Graphviz V 2.29.
	- Add data/58.gv, which uses a clickable href in a table as an edge label.
	- Replace sub read_cvs_file(), using IO::File and Text::CSV_XS, with Text::CSV::Slurp.

1.07  Thu Nov  8 12:38:00 2012
	- No code changes.
	- For pre-reqs such as strict, warnings, etc, which ship with Perl, set the version # to 0.
		Reported as RT#80663 by Father Chrysostomos for Tree::DAG_Node.

1.06  Fri Oct 26 09:09:00 2012
	- Change the lexer to output start_scope and end_scope rather than open_brace and close_brace.
		Change the parser to output start_scope and end_scope rather than start_graph and end_graph.
		This emphasises the case of the stand-alone {...}. Change the default renderer to match.
	- Increment the brace count when a '{' is encountered. This means brace counts (after the first 1
		at the start of the input) will be 1 higher than before, and when a '}' is encountered they will
		also be 1 higher. Hence they count down to 1, not 0 as they used to do.
	- Add an entry to the lexer's FAQ to clarify output from these 3: digraph {}, subgraph {} and {}.
	- Fix a lexer bug which, when given {a [] b}, would lose node b because [] was empty. See data/54.gv.
	- Add samples data/53.gv, data/54.gv and data/55.gv.
	- Add a corresponding entry in the parser's FAQ.
	- Add getters edges(), nodes(), style() and type(), which return most but not all of the information
		in the input *.gv file. These are basically for use by GraphViz2::Marpa::PathUtils.
		In particular, subgraph info is still missing.
	- Add option report_forest for new() and as a mutator. Calling new(report_forest => 1) prints all the
		information provided with those 4 new getters.
	- Reserve the paths() mutator for later implementation. The current code gets into an infinite loop
		sometimes, and so is commented out.
	- Expand the POD for the lexer and the parse giving details on input and output stream content.
	- Move the template for the demo from html/graphviz2.marpa.index.tx to htdocs/assets/templates/graphviz/marpa/
		where it should have been all along. Patch scripts/generate.demo.pl to match.
	- Add cache control metadata to all templates.
	- Link each template to the others, and to the index page for all graph module demos.
	- Add code to GraphViz2::Marpa::Utils to extract mutator names from modules and some programs,
		and output to data/code.attributes.csv, for use by scripts/code.attributes2html.pl.
	- Incorporate code from scripts/code.attributes2html.pl into GraphViz2::Marpa::Utils.
	- Incorporate code from scripts/stt2html.pl into GraphViz2::Marpa::Utils.
	- Rename html/default.stt.html to html/stt.html.
	- Expand contents of html/code.attributes.html, html/index.html and html/stt.html.
	- Add config/.htgraphviz2.marpa.conf and lib/GraphViz2/Marpa/Config.pm to simplify generating the demo's
		index.html.
	- Switch from Tree to Tree::DAG_Node, since it has various methods I hope to use implementing paths().

1.05  Tue Oct 16 09:49:00 2012
	- Fix a bug in the lexer where it used to combine multiple lines in the input file by, basically, using
		join('', <IN>), which meant the last char of a line was juxtaposed with the first char of the next
		line. If the second line started with spaces, that was ok, but if it started with, say, the name of
		a node, then that name was concatenated with what could have been a node name on the end of a previous
		line. This meant 2 separate node names sometimes became one.
	- Add data/52.gv to test the solution.
	- Shift data/code.attributes.html to html/.
	- Shift data/default.stt.html to html/.
	- Add scripts/generate.demo.sh.

1.04  Tue Sep 25 13:55:00 2012
	- All data/*.dot files were renamed to data/*.gv, in line with Graphviz itself.
	- Design change. In the output streams from both the lexer and the parser, the 3 special tokens whose names
		are edge, graph and node, were given the type node_id. Unfortunately, this is confusing because any
		normal edge, graph or node was also given that type. Now, these 3 cases are of type class_id.
		This makes it easier for code processing these output streams to distinguish between special cases
		and common-or-garden nodes. It should have always been done this way, and specifically simplifies code
		in GraphViz2::Marpa::TreeUtils.
	- Change Marpa.pm to support the new distro GraphViz2::Marpa::TreeUtils, which is a sub-class of GraphViz2::Marpa.

Makefile.PL
use Config;

use ExtUtils::MakeMaker;

use inc::Module::Install;

use File::Spec;
use File::Temp;

# -----------------------------------------------

sub is_dot_installed
{
	# 1: Create a temp file containing DOT commands.
	# The EXLOCK option is for BSD-based systems.
	# newdir() croaks() if it fails, which is what we want.
	# Hence there is no need to use Try::Tiny.

	my($temp_dir) = File::Temp -> newdir('temp.XXXX', CLEANUP => 1, EXLOCK => 0, TMPDIR => 1);
	my($gv_file)  = File::Spec -> catfile($temp_dir, 'test.gv');

	# We use the deprecated 2-argument form of open in case
	# this code is run on an ancient version of Perl.

	open(OUT, ">$gv_file")                   || die "Can't create temp file: $!\n";
	print OUT "digraph graph_14 {node_14}\n" || die "Can't write to temp file: $!\n";
	close OUT                                || die "Can't close temp file: $!\n";

	# 2: Run dot to create an SVG file.

	my $stdout = `dot -Tsvg $gv_file`;

	# 3: If that failed, we die.

	die "Please install Graphviz from http://www.graphviz.org/\n" if ($stdout !~ m|</svg>|);

} # End of is_dot_installed.

# -----------------------------------------------

sub is_perl_recent
{
	# We want Perl V 5.12 or later, because we 'use feature qw/say unicode_strings/'.

	my(@version) = split(/\./, $Config{version});

	die "Please install Perl V 5.12 or later\n" if ($version[1] < 12);

} # End of is_perl_recent.

# -----------------------------------------------

is_dot_installed();
is_perl_recent();

name 'GraphViz2-Marpa';

license 'artistic_2';

perl_version '5.014000';

all_from 'lib/GraphViz2/Marpa.pm';

install_share;

requires 'Algorithm::Diff'             => 1.1902;
requires 'Config'                      => undef;
requires 'Config::Tiny'                => 2.14;
requires 'Data::Section::Simple'       => 0.02;
requires 'Date::Format'                => 2.24;
requires 'Date::Simple'                => 3.03;
requires 'feature'                     => 0;
requires 'File::Spec'                  => 3.40;
requires 'File::ShareDir'              => 1.03;
requires 'File::Slurp'                 => 9999.19;
requires 'File::Temp'                  => 0.2301;
requires 'Getopt::Long'                => 2.38;
requires 'GraphViz2'                   => 1.09;
requires 'GraphViz2::Parse::Marpa'     => 1.12;
requires 'Hash::FieldHash'             => 0.10;
requires 'HTML::Entities::Interpolate' => 1.04;
requires 'List::Compare'               => 0.37;
requires 'Log::Handler'                => 0.71;
requires 'Marpa::R2'                   => 2.022000;
requires 'Module::Path'                => 0.08;
requires 'open'                        => 0;
requires 'Path::Tiny'                  => 0.022;
requires 'Perl6::Slurp'                => 0.03;
requires 'Pod::Usage'                  => 1.36;
requires 'Set::Array'                  => 0.23;
requires 'Set::FA::Element'            => 1.08;
requires 'strict'                      => 0;
requires 'Text::CSV'                   => 1.21;
requires 'Text::CSV::Slurp'            => 1;
requires 'Text::Xslate'                => 1.4001;
requires 'Tree::DAG_Node'              => 1.07;
requires 'Try::Tiny'                   => 0.09;
requires 'warnings'                    => 0;

test_requires 'Test::More' => 0.98;
#test_requires 'Test::Pod'  => 1.45; # Make it optional. See t/pod.t

WriteAll;
